### PR TITLE
Curate output

### LINKED ISSUE #142: Curate output
Right now we do not care much about the output. I wanted to pay a bit more attention to that.

I propose to keep the output limited to build backend output, this means swallowing the ensurepip output, and to curated messages from us (`build`). Our own messages should be clear that come from us, either by having an identifiable prefix or a specific style. Our messages should also be pretty limited, keeping the output clean. If we see need for debug messages we can offer that under a `--debug` flag.

For our messages, I propose adding a `build.print_info` method that could be overwritten, in `warnings.showwarning` fashion. The default implementation would use the `logging` module, but we would overwrite it in `__main__` to use a colorized raw print.

### PR DESCRIPTION
**Pull request by FFY00**
Fixes #142

### CODE REVIEW

**Comment by FFY00** [comment]
Preview: https://asciinema.org/a/tjnnnrs94SF3ZdOmYvnOsXzxp

File: src/build/env.py
Comment (gaborbernat) [inline src/build/env.py:85]: This catch seems to do nothing :thinking:

File: src/build/env.py
Comment (gaborbernat) [inline src/build/env.py:75]: args needs to be a sequence of str, not? Also for kwargs can we just state other parameters we use?

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: src/build/env.py
Comment (layday) [inline src/build/env.py:195]: on -> in

File: src/build/env.py
Comment (layday) [reply src/build/env.py:85]: Don't you just want `subprocess.check_output` here?

Comment (FFY00) [reply src/build/env.py:85]: Yes, sorry. It accidentally got removed.

File: src/build/env.py
Comment (gaborbernat) [inline src/build/env.py:80]: perhaps open an issue to typeshed and link it here instead?

File: src/build/env.py
Comment (gaborbernat) [inline src/build/env.py:79]: this should be just subprocess.ProcessCaledError, not?

Comment (gaborbernat) [reply src/build/env.py:79]: Can we also create a test for this? :thinking:

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: src/build/env.py
Comment (layday) [reply src/build/env.py:81]: You're catching a generic `Exception`, that's why it's not type hinted. Catch `CalledProcessError` and let everything else through.

Comment (FFY00) [reply src/build/env.py:81]: Yes, of course, sorry. I just woke up :sweat_smile:

File: src/build/__main__.py
Comment (gaborbernat) [inline src/build/__main__.py:38]: Why does this require pragma no cover?

File: src/build/env.py
Comment (gaborbernat) [inline src/build/env.py:79]: Why does this require pragma no cover? Can we add a test to not need it?

**Comment by FFY00** [comment]
Sorry about my previous fix, I had just woken up and was sleepy :sweat_smile:. I fixed this properly now and also added dim to the tracebacks.

File: src/build/env.py
Comment (FFY00) [reply src/build/env.py]: Like pointed out above, I am catching `Exception`, which will obviously not have `e.output` :sweat_smile:

**Review by gaborbernat — APPROVED** [review]

File: src/build/__main__.py
Comment (layday) [inline src/build/__main__.py:41]: Why not just use string formatting instead of ad hoc replaces?  Something like:

```py
def _print(message: str) -> None:  # pragma: no cover
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    if color:
        message = message.format(**_STYLES)
    else:
        message = message.format(**{k: '' for k in _STYLES})
    print(message)

[...]

_print(f'{{yellow}}WARNING{{reset}} {message}')
```

Comment (FFY00) [reply src/build/__main__.py:41]: The performance difference is not relevant, and I felt this had slightly cleaner code and considerably better readability in the messages -- the color formatting is clearly distinct from f-string variables and only takes one delimiter character on each side.

```python
import os
import sys


_STYLES = {
    'a': '1',
    'b': '2',
    'c': '3',
    'd': '4',
    'e': '5',
}


def _print_replace(message: str) -> None:
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    for name, code in _STYLES.items():
        message = message.replace(f'[{name}]', code if color else '')
    #print(message)


def _print_format(message: str) -> None:  # pragma: no cover
    color = sys.stdout.isatty() and 'NO_COLOR' not in os.environ
    if color:
        message = message.format(**_STYLES)
    else:
        message = message.format(**{k: '' for k in _STYLES})
    #print(message)
```

```
$ python -m timeit -s 'import test_print' -- 'test_print._print_replace("[a][b][c][d][e]")'
100000 loops, best of 5: 2.35 usec per loop
$ python -m timeit -s 'import test_print' -- 'test_print._print_format("{{a}}{{b}}{{c}}{{d}}{{e}}")'
200000 loops, best of 5: 1.93 usec per loop
```

Comment (layday) [reply src/build/__main__.py:41]: I don't think the performance difference is important either - I just felt it's the obvious way to do string substitution. Let's keep the squares then if you think readability would suffer.

**Comment by FFY00** [comment]
Hum, I though the automerge would wait for @henryiii as I requested a review from him. Let me know if there are any issues.

**Comment by layday** [comment]
It didn't wait for me to approve either but it's all good 😛

**Comment by FFY00** [comment]
Because you commented it removed the review request anyway :smile:

File: src/build/__main__.py
Comment (henryiii) [reply src/build/__main__.py:41]: This looks a lot like rich's mini language - why not exactly imitate that? `[yellow]<stuff>[/yellow]`. You could loop over over styles and add `/` versions that reset. Just a thought, instead of inventing something new just for this. _Very_ minor though.

Comment (henryiii) [reply src/build/__main__.py:41]: Also what happens if the string `[yellow]` happens to be part of the traceback, since tracebacks are passed through this? Doing fstring replacement wouldn't have this problem.

Comment (FFY00) [reply src/build/__main__.py:41]: More complexity in the implementation :stuck_out_tongue:. This is private anyway, so I just went with the simple thing that would fit our needs.

File: src/build/__main__.py
Comment (henryiii) [inline src/build/__main__.py:157]: Here's where it looks like `[<stuff>]` could get introduces in the tb and miss-processed.

File: src/build/__main__.py
Comment (FFY00) [reply src/build/__main__.py:41]: > Also what happens if the string `[yellow]` happens to be part of the traceback, since tracebacks are passed through this? Doing fstring replacement wouldn't have this problem.

Right, we should not pass them through.

Comment (henryiii) [reply src/build/__main__.py:41]: Personally, in plumbum.colorlib, I use `print(yellow | "str") and `yellow` wraps the string it's or'd with with color tags.

Comment (FFY00) [reply src/build/__main__.py:41]: Right, but that introduces a bit more complexity than what we need. I opened #335.

**Pull request merged**
